File: hw2/cdc/cdc.cpp
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

#define WIN_SIZE 16
#define PRIME 3
#define MODULUS 256
#define TARGET 0

uint64_t hash_func(unsigned char *input, unsigned int pos)
{
	// put your hash function implementation here
}

void cdc(unsigned char *buff, unsigned int buff_size)
{
	// put your cdc implementation here

}

void test_cdc( const char* file )
{
	FILE* fp = fopen(file,"r" );
	if(fp == NULL ){
		perror("fopen error");
		return;
	}

	fseek(fp, 0, SEEK_END); // seek to end of file
	int file_size = ftell(fp); // get current file pointer
	fseek(fp, 0, SEEK_SET); // seek back to beginning of file

	unsigned char* buff = (unsigned char *)malloc((sizeof(unsigned char) * file_size ));	
	if(buff == NULL)
	{
		perror("not enough space");
		fclose(fp);
		return;
	}

	int bytes_read = fread(&buff[0],sizeof(unsigned char),file_size,fp);

	cdc(buff, file_size);

    free(buff);
    return;
}

int main()
{
	test_cdc("prince.txt");
	return 0;
}

File: hw2/cdc/cdc.py
win_size = 16
prime = 3
modulus = 256
target = 0


def hash_func(input, pos):
    hash = 0
    for i in range(0, win_size):
        hash += ord(input[pos+win_size-1-i])*(pow(prime, i+1))
    return hash


def cdc(buff, buff_size):
    for i in range(win_size, buff_size-win_size):
        if((hash_func(buff, i) % modulus)) == target:
            print(i)


def test_cdc(filename):
    with open(filename) as f:
        buff = f.read()
        cdc(buff, len(buff))


test_cdc("prince.txt")
